modules/python-modules/syslogng/modules/s3/s3_destination.py

```diff
@@ -0,0 +1,444 @@
+
+        for finished_timestamps in self.finished_s3_objects.values():
+            for indexes in finished_timestamps.values():
+                for s3_object in indexes.values():
+                    s3_object.wait_for_upload_to_complete()
+
+        self.finished_s3_objects.clear()
+
+        self.probably_connected = False
+
+        self.client.close()
+        self.client = None
+
+    def __start_flush_poll_timer(self) -> None:
+        self.flush_poll_event = Timer(
+            interval=self.S3_OBJECT_TIMEOUT_INTERVAL_SECONDS, function=self.__flush_timed_out_s3_objects
+        )
+        self.flush_poll_event.start()
+
+    def __flush_timed_out_s3_objects(self) -> None:
+        def should_flush_s3_object(s3_object: S3Object, now: float) -> bool:
+            return not s3_object.finished and s3_object.modified_at + self.flush_grace_period * 60 <= now
+
+        now = monotonic()
+
+        for s3_object in self.s3_objects.values():
+            if should_flush_s3_object(s3_object, now):
+                self.__finish_s3_object(s3_object)
+
+        for timestamps in self.backfill_s3_objects.values():
+            for s3_object in timestamps.values():
+                if should_flush_s3_object(s3_object, now):
+                    self.__finish_s3_object(s3_object)
+
+        self.__start_flush_poll_timer()
+
+    def __format_target_key(self, msg: LogMessage) -> str:
+        return self.object_key.format(msg, self.template_options)
+
+    def __format_timestamp(self, msg: LogMessage) -> str:
+        if self.object_key_timestamp is not None:
+            return self.object_key_timestamp.format(msg, self.template_options)
+        return ""
+
+    def __create_initial_s3_object(self, target_key: str, timestamp: str) -> S3Object:
+        return S3Object.create_initial(
+            working_dir=self.working_dir,
+            bucket=self.bucket,
+            target_key=target_key,
+            timestamp=timestamp,
+            compress=self.compression,
+            storage_class=self.storage_class,
+            persist_name=self.persist_name,
+            executor=self.executor,
+            compresslevel=self.compresslevel,
+            chunk_size=self.chunk_size,
+            canned_acl=self.canned_acl,
+            client=self.client,
+            logger=self.logger,
+            exit_requested=self.exit_requested,
+        )
+
+    def __finish_s3_object(self, s3_object: S3Object) -> None:
+        if s3_object.finished:
+            return
+
+        s3_object.finish()
+        d = self.finished_s3_objects.setdefault(s3_object.target_key, dict()).setdefault(s3_object.timestamp, dict())
+        d[s3_object.index] = s3_object
+
+    def __get_backfill_s3_object(self, target_key: str, timestamp: str) -> S3Object:
+        backfill_target_s3_objects = self.backfill_s3_objects.setdefault(target_key, dict())
+
+        # Completely new backfill
+        if timestamp not in backfill_target_s3_objects.keys():
```

Comment:
In python3 the `keys()` call no longer returns a list, but a `dict_keys` object, which has an O(1) lookup time complexity.

However it is recommended to just use `key in my_dict`, because it is simpler.

https://wiki.python.org/moin/TimeComplexity

Reply:
Fixed.

Reply:
Indeed, sorry.